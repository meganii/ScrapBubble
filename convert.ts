import { toTitleLc } from "./deps/scrapbox-std.ts";
import { Bubble, BubbleStorage } from "./storage.ts";
import { fromId, ID, toId } from "./id.ts";
import { getUnixTime } from "./time.ts";
import { Page, RelatedPage } from "./deps/scrapbox.ts";

/** APIから取得したページデータを、Bubble用に変換する
 *
 * @param titleLc ページのタイトル タイトル変更があると、page.titleから復元できないため、別途指定している
 * @param project ページのproject
 * @param page 変換したいページデータ
 * @return 変換したデータ
 */
export const convert = (
  project: string,
  page: Page,
  checked?: Date,
): BubbleStorage => {
  const storage: BubbleStorage = new Map();
  checked ??= new Date();

  // pageが参照しているリンクの逆リンクにpageを入れる
  // これにより、2 hop linksのハブとなるcardは、全ての逆リンクが格納されていると保証され、linkedとprojectLinkedはundefinedでなくなる
  const titleLc = toTitleLc(page.title);
  for (const link of page.links) {
    const bubble: Bubble = makeDummy(project, link, checked);
    bubble.linked = [titleLc];
    storage.set(toId(project, link), bubble);
  }
  const pageId = toId(project, titleLc);
  const projectLinksLc: ID[] = [];
  for (const id of page.projectLinks) {
    // idは/:project/:title という形式だと保証していい
    const link = fromId(id as ID);
    const cardId = toId(link.project, link.titleLc);
    projectLinksLc.push(cardId);
    // `link.titleLc`はtitle形式となる
    const bubble: Bubble = makeDummy(link.project, link.titleLc, checked);
    bubble.projectLinked = [pageId];
    storage.set(cardId, bubble);
  }

  // ページ本文を入れる
  const pageBubble: Required<Bubble> = {
    ...toBubble(project, page, checked),
    linked: [],
    projectLinked: [],
  };
  storage.set(pageId, pageBubble);

  // 1 hop linksからカードを取り出す
  // 同時に`emptyPageIds`を作る
  for (const card of page.relatedPages.links1hop) {
    if (card.linksLc.includes(titleLc)) {
      // 双方向リンク or 逆リンク

      // 逆リンクを作る
      // 重複はありえないので配列でいい
      pageBubble.linked.push(card.titleLc);
    }
    // cardを入れる
    const cardId = toId(project, card.titleLc);
    const bubble = toBubble(project, card, checked);
    // external linksでなければ`projectLinked`は存在しない
    const linked = storage.get(cardId)?.linked;
    if (linked) bubble.linked = linked;
    storage.set(cardId, bubble);
  }

  // external linksからカードを取り出す
  for (
    const card of page.relatedPages.projectLinks1hop
  ) {
    const cardId = toId(card.projectName, card.titleLc);
    if (!projectLinksLc.includes(cardId)) {
      // 逆リンク
      // 双方向リンクは順リンクと判別がつかないのでやめる

      // 逆リンクを作る
      pageBubble.projectLinked.push(cardId);
    }
    // cardを入れる
    const bubble = toBubble(card.projectName, card, checked);
    const projectLinked = storage.get(cardId)?.projectLinked;
    if (projectLinked) bubble.projectLinked = projectLinked;
    storage.set(cardId, bubble);
  }

  // 2 hop linksからカードを取り出す
  for (const card of page.relatedPages.links2hop) {
    for (const linkLc of card.linksLc) {
      // 逆リンクを作る
      // 重複はありえないので配列でいい
      const cardId = toId(project, linkLc);
      const bubble = storage.get(cardId);
      if (!bubble) throw Error(`storage already must have "${cardId}"`);
      if (!bubble.linked) {
        bubble.linked = [card.titleLc];
        continue;
      }
      bubble.linked.push(card.titleLc);
    }
    // cardを入れる
    const cardId = toId(project, card.titleLc);
    const bubble = toBubble(project, card, checked);
    // external linksでなければ`projectLinked`は存在しない
    const linked = storage.get(cardId)?.linked;
    if (linked) bubble.linked = linked;
    storage.set(cardId, bubble);
  }

  return storage;
};

/** 関連ページやページ本文の情報をBubbleに変換する
 *
 * @param project 与えたデータのproject name
 * @param page bubbleに変換したいデータ
 * @param checked データを確認した日時
 * @return 変換後のデータ
 */
const toBubble = (
  project: string,
  page: Page | Omit<RelatedPage, "linksLc">,
  checked: Date,
): Bubble => ({
  project,
  titleLc: "titleLc" in page ? page.titleLc : toTitleLc(page.title),
  // 関連ページの場合は、関連ページが存在する時点で中身があるので、常に`true`とする
  exists: "persistent" in page ? page.persistent : true,
  descriptions: page.descriptions,
  image: page.image,
  lines: "lines" in page
    ? page.lines
    // descriptionsからでっち上げる
    : [page.title, ...page.descriptions].map((text) => ({
      text,
      id: "dummy",
      userId: "dummy",
      updated: page.updated,
      created: page.updated,
    })),
  updated: page.updated,
  checked: getUnixTime(checked),
});

/** ページタイトルだけからBubbleを作る
 *
 * @param project 与えたデータのproject name
 * @param title ページタイトル
 * @param checked データを確認した日時
 * @return 変換後のデータ
 */
const makeDummy = (
  project: string,
  title: string,
  checked: Date,
): Bubble => ({
  project,
  titleLc: toTitleLc(title),
  exists: false,
  descriptions: [],
  image: null,
  lines: [{
    text: title,
    id: "dummy",
    userId: "dummy",
    updated: 0,
    created: 0,
  }],
  updated: 0,
  checked: getUnixTime(checked),
});
