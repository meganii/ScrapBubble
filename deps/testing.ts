export * from "https://deno.land/std@0.174.0/testing/asserts.ts";
export * from "https://deno.land/std@0.174.0/testing/snapshot.ts";
