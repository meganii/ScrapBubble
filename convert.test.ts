import { convert } from "./convert.ts";
import { assertSnapshot } from "./deps/testing.ts";

Deno.test("convert", async (t) => {
  await t.step("one project", async (t) => {
    await t.step("たぶん全パターン入り", async (t) => {
      await assertSnapshot(
        t,
        convert("takker-dist", {
          "id": "63ab57239db1b2001edc43b0",
          "title": "B",
          "image": null,
          "descriptions": [
            "[B1], [B2], [B3]",
            "[A1]",
            "https://scrapbox.io/api/pages/takker-dist/B",
          ],
          "user": {
            "id": "5ef2bdebb60650001e1280f0",
            "name": "takker",
            "displayName": "takker",
            "photo":
              "https://lh3.googleusercontent.com/a/AEdFTp5ZjA62crJpg6_eUg0wkIvPXm36hWWC3gyzzuZZ=s96-c",
          },
          "pin": 0,
          "views": 5,
          "linked": 1,
          "commitId": "63b93444aded4c001e09d134",
          "created": 1672173349,
          "updated": 1673081924,
          "accessed": 1673257686,
          "snapshotCreated": 1673081924,
          "snapshotCount": 1,
          "pageRank": 1.8,
          "lastAccessed": 1673257686,
          "persistent": true,
          "lines": [{
            "id": "63ab57239db1b2001edc43b0",
            "text": "B",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1672173349,
            "updated": 1672173349,
          }, {
            "id": "63ab57261280f000002d941d",
            "text": "[B1], [B2], [B3]",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1672173349,
            "updated": 1672173361,
          }, {
            "id": "63ab575f1280f000002d941f",
            "text": "[A1]",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1672173407,
            "updated": 1672173410,
          }, {
            "id": "63ab57271280f000002d941e",
            "text": "https://scrapbox.io/api/pages/takker-dist/B",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1672173351,
            "updated": 1673081924,
          }, {
            "id": "63b934441280f00000606f45",
            "text": "",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673081924,
            "updated": 1673081924,
          }],
          "links": ["B1", "B2", "B3", "A1"],
          "projectLinks": [],
          "icons": [],
          "files": [],
          "relatedPages": {
            "links1hop": [{
              "id": "63b933f82895b8001d9f5cd1",
              "title": "B1",
              "titleLc": "b1",
              "image": null,
              "descriptions": ["[B]"],
              "linksLc": ["b"],
              "linked": 4,
              "updated": 1673081852,
              "accessed": 1673082412,
            }],
            "links2hop": [{
              "id": "63ab5703d28d9b001d611a14",
              "title": "A",
              "titleLc": "a",
              "image": null,
              "descriptions": [
                "[A1],[A2],[A3]",
                "https://scrapbox.io/api/pages/takker-dist/A",
              ],
              "linksLc": ["a1"],
              "linked": 0,
              "updated": 1672173548,
              "accessed": 1673081915,
            }, {
              "id": "63ab57bd101f5f001ee7125c",
              "title": "C",
              "titleLc": "c",
              "image": null,
              "descriptions": ["[C1], [C2], [C3]", "[A2], [A3]", "[B1]"],
              "linksLc": ["b1"],
              "linked": 0,
              "updated": 1673081866,
              "accessed": 1673081862,
            }, {
              "id": "63b93413cb7e16001e1143af",
              "title": "D",
              "titleLc": "d",
              "image": null,
              "descriptions": ["[B1]"],
              "linksLc": ["b1"],
              "linked": 0,
              "updated": 1673081879,
              "accessed": 1673081876,
            }, {
              "id": "63b9343133a85a001ea1d3c3",
              "title": "E",
              "titleLc": "e",
              "image": null,
              "descriptions": ["[B1], [B2]"],
              "linksLc": ["b1", "b2"],
              "linked": 0,
              "updated": 1673081909,
              "accessed": 1673081906,
            }],
            "projectLinks1hop": [],
            "hasBackLinksOrIcons": true,
          },
          "collaborators": [],
        }, new Date(1672173548000)),
      );
    });
  });
  await t.step("two projects", async (t) => {
    await t.step("逆・順・双方向External links", async (t) => {
      await assertSnapshot(
        t,
        convert("takker-dist", {
          "id": "63bd16afdb71fb001d058f0e",
          "title": "F",
          "image": null,
          "descriptions": [
            "[/takker-dist2/B]",
            "[/takker-dist2/C]",
            "[/takker-dist2/G]",
            "https://scrapbox.io/api/pages/takker-dist/F",
          ],
          "user": {
            "id": "5ef2bdebb60650001e1280f0",
            "name": "takker",
            "displayName": "takker",
            "photo":
              "https://lh3.googleusercontent.com/a/AEdFTp5ZjA62crJpg6_eUg0wkIvPXm36hWWC3gyzzuZZ=s96-c",
          },
          "pin": 0,
          "views": 3,
          "linked": 0,
          "commitId": "63bd174061a69e001e7d9d74",
          "created": 1673336497,
          "updated": 1673336640,
          "accessed": 1673336624,
          "snapshotCreated": null,
          "snapshotCount": 0,
          "pageRank": 0,
          "lastAccessed": 1673336624,
          "persistent": true,
          "lines": [{
            "id": "63bd16afdb71fb001d058f0e",
            "text": "F",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673336497,
            "updated": 1673336497,
          }, {
            "id": "63bd16b11280f00000b00965",
            "text": "[/takker-dist2/B]",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673336497,
            "updated": 1673336501,
          }, {
            "id": "63bd16b51280f00000b00966",
            "text": "[/takker-dist2/C]",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673336501,
            "updated": 1673336571,
          }, {
            "id": "63bd17041280f00000b0096b",
            "text": "[/takker-dist2/G]",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673336580,
            "updated": 1673336595,
          }, {
            "id": "63bd17111280f00000b0096c",
            "text": "https://scrapbox.io/api/pages/takker-dist/F",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673336594,
            "updated": 1673336640,
          }, {
            "id": "63bd17401280f00000b0096e",
            "text": "",
            "userId": "5ef2bdebb60650001e1280f0",
            "created": 1673336640,
            "updated": 1673336640,
          }],
          "links": [],
          "projectLinks": [
            "/takker-dist2/B",
            "/takker-dist2/C",
            "/takker-dist2/G",
          ],
          "icons": [],
          "files": [],
          "relatedPages": {
            "links1hop": [],
            "links2hop": [],
            "projectLinks1hop": [{
              "id": "63bd169823cbed001d092543",
              "title": "B",
              "titleLc": "b",
              "image": null,
              "descriptions": ["[B1] [B2]"],
              "linked": 0,
              "updated": 1673336514,
              "accessed": 1673336473,
              "projectName": "takker-dist2",
            }, {
              "id": "63bd16c41933fc001d0f9dca",
              "title": "C",
              "titleLc": "c",
              "image": null,
              "descriptions": ["[/takker-dist/F]"],
              "linked": 0,
              "updated": 1673336584,
              "accessed": 1673336577,
              "projectName": "takker-dist2",
            }, {
              "id": "63bd1684ca5336001d6417d9",
              "title": "A",
              "titleLc": "a",
              "image": null,
              "descriptions": ["[/takker-dist/F]"],
              "linked": 0,
              "updated": 1673336561,
              "accessed": 1673336529,
              "projectName": "takker-dist2",
            }],
            "hasBackLinksOrIcons": false,
          },
          "collaborators": [],
        }, new Date(1672173548000)),
      );
    });
  });
});
