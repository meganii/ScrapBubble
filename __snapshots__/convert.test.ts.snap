export const snapshot = {};

snapshot[`convert > one project > たぶん全パターン入り 1`] = `
Map {
  "/takker-dist/a" => {
      checked: 1672173549,
      descriptions: [
        "[A1],[A2],[A3]",
        "https://scrapbox.io/api/pages/takker-dist/A",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1672173548,
          id: "dummy",
          text: "A",
          updated: 1672173548,
          userId: "dummy",
        },
        {
          created: 1672173548,
          id: "dummy",
          text: "[A1],[A2],[A3]",
          updated: 1672173548,
          userId: "dummy",
        },
        {
          created: 1672173548,
          id: "dummy",
          text: "https://scrapbox.io/api/pages/takker-dist/A",
          updated: 1672173548,
          userId: "dummy",
        },
      ],
      project: "takker-dist",
      titleLc: "a",
      updated: 1672173548,
    },
  "/takker-dist/a1" => {
      checked: 1672173549,
      descriptions: [
      ],
      exists: false,
      image: null,
      lines: [
        {
          created: 0,
          id: "dummy",
          text: "A1",
          updated: 0,
          userId: "dummy",
        },
      ],
      linked: [
        "b",
        "a",
      ],
      project: "takker-dist",
      titleLc: "a1",
      updated: 0,
    },
  "/takker-dist/b" => {
      checked: 1672173549,
      descriptions: [
        "[B1], [B2], [B3]",
        "[A1]",
        "https://scrapbox.io/api/pages/takker-dist/B",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1672173349,
          id: "63ab57239db1b2001edc43b0",
          text: "B",
          updated: 1672173349,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1672173349,
          id: "63ab57261280f000002d941d",
          text: "[B1], [B2], [B3]",
          updated: 1672173361,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1672173407,
          id: "63ab575f1280f000002d941f",
          text: "[A1]",
          updated: 1672173410,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1672173351,
          id: "63ab57271280f000002d941e",
          text: "https://scrapbox.io/api/pages/takker-dist/B",
          updated: 1673081924,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1673081924,
          id: "63b934441280f00000606f45",
          text: "",
          updated: 1673081924,
          userId: "5ef2bdebb60650001e1280f0",
        },
      ],
      linked: [
        "b1",
      ],
      project: "takker-dist",
      projectLinked: [
      ],
      titleLc: "b",
      updated: 1673081924,
    },
  "/takker-dist/b1" => {
      checked: 1672173549,
      descriptions: [
        "[B]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673081852,
          id: "dummy",
          text: "B1",
          updated: 1673081852,
          userId: "dummy",
        },
        {
          created: 1673081852,
          id: "dummy",
          text: "[B]",
          updated: 1673081852,
          userId: "dummy",
        },
      ],
      linked: [
        "b",
        "c",
        "d",
        "e",
      ],
      project: "takker-dist",
      titleLc: "b1",
      updated: 1673081852,
    },
  "/takker-dist/b2" => {
      checked: 1672173549,
      descriptions: [
      ],
      exists: false,
      image: null,
      lines: [
        {
          created: 0,
          id: "dummy",
          text: "B2",
          updated: 0,
          userId: "dummy",
        },
      ],
      linked: [
        "b",
        "e",
      ],
      project: "takker-dist",
      titleLc: "b2",
      updated: 0,
    },
  "/takker-dist/b3" => {
      checked: 1672173549,
      descriptions: [
      ],
      exists: false,
      image: null,
      lines: [
        {
          created: 0,
          id: "dummy",
          text: "B3",
          updated: 0,
          userId: "dummy",
        },
      ],
      linked: [
        "b",
      ],
      project: "takker-dist",
      titleLc: "b3",
      updated: 0,
    },
  "/takker-dist/c" => {
      checked: 1672173549,
      descriptions: [
        "[C1], [C2], [C3]",
        "[A2], [A3]",
        "[B1]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673081866,
          id: "dummy",
          text: "C",
          updated: 1673081866,
          userId: "dummy",
        },
        {
          created: 1673081866,
          id: "dummy",
          text: "[C1], [C2], [C3]",
          updated: 1673081866,
          userId: "dummy",
        },
        {
          created: 1673081866,
          id: "dummy",
          text: "[A2], [A3]",
          updated: 1673081866,
          userId: "dummy",
        },
        {
          created: 1673081866,
          id: "dummy",
          text: "[B1]",
          updated: 1673081866,
          userId: "dummy",
        },
      ],
      project: "takker-dist",
      titleLc: "c",
      updated: 1673081866,
    },
  "/takker-dist/d" => {
      checked: 1672173549,
      descriptions: [
        "[B1]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673081879,
          id: "dummy",
          text: "D",
          updated: 1673081879,
          userId: "dummy",
        },
        {
          created: 1673081879,
          id: "dummy",
          text: "[B1]",
          updated: 1673081879,
          userId: "dummy",
        },
      ],
      project: "takker-dist",
      titleLc: "d",
      updated: 1673081879,
    },
  "/takker-dist/e" => {
      checked: 1672173549,
      descriptions: [
        "[B1], [B2]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673081909,
          id: "dummy",
          text: "E",
          updated: 1673081909,
          userId: "dummy",
        },
        {
          created: 1673081909,
          id: "dummy",
          text: "[B1], [B2]",
          updated: 1673081909,
          userId: "dummy",
        },
      ],
      project: "takker-dist",
      titleLc: "e",
      updated: 1673081909,
    },
}
`;

snapshot[`convert > two projects > 逆・順・双方向External links 1`] = `
Map {
  "/takker-dist/f" => {
      checked: 1672173548,
      descriptions: [
        "[/takker-dist2/B]",
        "[/takker-dist2/C]",
        "[/takker-dist2/G]",
        "https://scrapbox.io/api/pages/takker-dist/F",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673336497,
          id: "63bd16afdb71fb001d058f0e",
          text: "F",
          updated: 1673336497,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1673336497,
          id: "63bd16b11280f00000b00965",
          text: "[/takker-dist2/B]",
          updated: 1673336501,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1673336501,
          id: "63bd16b51280f00000b00966",
          text: "[/takker-dist2/C]",
          updated: 1673336571,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1673336580,
          id: "63bd17041280f00000b0096b",
          text: "[/takker-dist2/G]",
          updated: 1673336595,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1673336594,
          id: "63bd17111280f00000b0096c",
          text: "https://scrapbox.io/api/pages/takker-dist/F",
          updated: 1673336640,
          userId: "5ef2bdebb60650001e1280f0",
        },
        {
          created: 1673336640,
          id: "63bd17401280f00000b0096e",
          text: "",
          updated: 1673336640,
          userId: "5ef2bdebb60650001e1280f0",
        },
      ],
      linked: [
      ],
      project: "takker-dist",
      projectLinked: [
        "/takker-dist2/a",
      ],
      titleLc: "f",
      updated: 1673336640,
    },
  "/takker-dist2/a" => {
      checked: 1672173548,
      descriptions: [
        "[/takker-dist/F]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673336561,
          id: "dummy",
          text: "A",
          updated: 1673336561,
          userId: "dummy",
        },
        {
          created: 1673336561,
          id: "dummy",
          text: "[/takker-dist/F]",
          updated: 1673336561,
          userId: "dummy",
        },
      ],
      project: "takker-dist2",
      titleLc: "a",
      updated: 1673336561,
    },
  "/takker-dist2/b" => {
      checked: 1672173548,
      descriptions: [
        "[B1] [B2]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673336514,
          id: "dummy",
          text: "B",
          updated: 1673336514,
          userId: "dummy",
        },
        {
          created: 1673336514,
          id: "dummy",
          text: "[B1] [B2]",
          updated: 1673336514,
          userId: "dummy",
        },
      ],
      project: "takker-dist2",
      titleLc: "b",
      updated: 1673336514,
    },
  "/takker-dist2/c" => {
      checked: 1672173548,
      descriptions: [
        "[/takker-dist/F]",
      ],
      exists: true,
      image: null,
      lines: [
        {
          created: 1673336584,
          id: "dummy",
          text: "C",
          updated: 1673336584,
          userId: "dummy",
        },
        {
          created: 1673336584,
          id: "dummy",
          text: "[/takker-dist/F]",
          updated: 1673336584,
          userId: "dummy",
        },
      ],
      project: "takker-dist2",
      titleLc: "c",
      updated: 1673336584,
    },
}
`;
